tree-sitter-stack-graphs/src/functions.rs

```diff
@@ -0,0 +1,194 @@
+// -*- coding: utf-8 -*-
+// ------------------------------------------------------------------------------------------------
+// Copyright © 2021, stack-graphs authors.
+// Licensed under either of Apache License, Version 2.0, or MIT license, at your option.
+// Please see the LICENSE-APACHE or LICENSE-MIT files in this distribution for license details.
+// ------------------------------------------------------------------------------------------------
+
+//! Define tree-sitter-graph functions
+
+use tree_sitter_graph::functions::Functions;
+
+pub fn add_path_functions(functions: &mut Functions) {
+    functions.add("path-dir".into(), path::PathDir);
+    functions.add("path-fileext".into(), path::PathFileExt);
+    functions.add("path-filename".into(), path::PathFileName);
+    functions.add("path-filestem".into(), path::PathFileStem);
+    functions.add("path-join".into(), path::PathJoin);
+    functions.add("path-normalize".into(), path::PathNormalize);
+    functions.add("path-split".into(), path::PathSplit);
+}
+
+pub mod path {
+    use std::path::Component;
+    use std::path::PathBuf;
+    use tree_sitter_graph::functions::Function;
+    use tree_sitter_graph::functions::Parameters;
+    use tree_sitter_graph::graph::Graph;
+    use tree_sitter_graph::graph::Value;
+    use tree_sitter_graph::ExecutionError;
+
+    pub struct PathDir;
+    pub struct PathFileExt;
+    pub struct PathFileName;
+    pub struct PathFileStem;
+    pub struct PathJoin;
+    pub struct PathNormalize;
+    pub struct PathSplit;
+
+    impl Function for PathDir {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let path = PathBuf::from(parameters.param()?.into_string()?);
+            parameters.finish()?;
+
+            let path = path.parent();
+            Ok(path
+                .map(|p| p.to_str().unwrap().into())
+                .unwrap_or(Value::Null))
+        }
+    }
```

Comment:
Is it worth trying to abstract out the common part of this, i.e. all but `let path = …`? Or does it end up being approx. the same amount to understand for each call site?

Reply:
Good suggestion. Turns out this is indeed possible, and does clean up the code a bit!